### PR TITLE
FIX: Ensure dataobj datatype is altered

### PR DESCRIPTION
**Pull request by mgxd**
brought to attention in https://github.com/poldracklab/fmriprep/issues/2142

### CODE REVIEW

**Comment by effigies** [comment]
This will have no effect. `img.set_data_dtype()` is just a proxy:

https://github.com/nipy/nibabel/blob/966a01e63fd7ce7e58d5971184e31b3305d350f6/nibabel/analyze.py#L931-L932

**Comment by effigies** [comment]
I don't fully understand the `data_dtype` parameter. Is that strictly for the on-disk format? Or is it supposed to coerce the data array, as well?

**Comment by mgxd** [comment]
It was meant to only effect the on-disk format. I'm a bit hesitant to coerce the data within the datasink interface, as there is already a lot going on there. If we want to go that route, I'd feel more comfortable splitting that into its own Node

File: niworkflows/interfaces/bids.py [hunk 1]
```diff
@@ -596,8 +596,16 @@ def _run_interface(self, runtime):
...
                         nii = nb.load(out_file, mmap=False)
                     data_dtype = np.dtype(data_dtype)
-                    if nii.get_data_dtype() != data_dtype:
+                    orig_dtype = nii.get_data_dtype()
+                    if orig_dtype != data_dtype:
+                        LOGGER.warning(
+                            f"Changing {out_file} dtype from {orig_dtype} to {data_dtype}"
+                        )
+                        # coerce dataobj to new data dtype
+                        new_data = np.asanyarray(nii.dataobj).astype(data_dtype)
```
Comment (oesteban) [inline niworkflows/interfaces/bids.py:605]:
```suggestion
                        new_data = np.asanyarray(nii.dataobj, dtype=data_dtype)
```

Comment (effigies) [reply niworkflows/interfaces/bids.py:605]: Is this going to cast BOLD data to int16?

Comment (mgxd) [reply niworkflows/interfaces/bids.py:605]: i think https://github.com/poldracklab/smriprep/issues/195 is part of the problem, but there is some further step downstream that is upcasting to float. But this is also necessary for any conversion to match `source_file` (currently just `boldref`).

Comment (mgxd) [reply niworkflows/interfaces/bids.py]: @effigies this should only affect data suffixes:
```
        ("mask", "uint8"),
        ("dseg", "int16"),
        ("probseg", "float32"),
        ("boldref", "source")
```
unless the `data_dtype` flag is explicitly checked, which I don't believe it ever is.

Comment (oesteban) [reply niworkflows/interfaces/bids.py]: > i think [poldracklab/smriprep#195](https://github.com/poldracklab/smriprep/issues/195) is part of the problem, but there is some further step downstream that is upcasting to float.

commented there, yes there is a resampling downstream in this case - moving to standard space.

File: niworkflows/interfaces/bids.py [hunk 2]
```diff
@@ -596,8 +596,16 @@ def _run_interface(self, runtime):
...
                         nii = nb.load(out_file, mmap=False)
                     data_dtype = np.dtype(data_dtype)
-                    if nii.get_data_dtype() != data_dtype:
+                    orig_dtype = nii.get_data_dtype()
+                    if orig_dtype != data_dtype:
+                        LOGGER.warning(
+                            f"Changing {out_file} dtype from {orig_dtype} to {data_dtype}"
+                        )
+                        # coerce dataobj to new data dtype
+                        new_data = np.asanyarray(nii.dataobj, dtype=data_dtype)
```
Comment (effigies) [inline niworkflows/interfaces/bids.py:605]: Can't just cast to int without rounding.

```Python
>>> np.asanyarray(50.99999, np.int16)
array(50, dtype=int16)
```

What about:
```suggestion
                        if issubclass(data_dtype, np.integer):
                            new_data = np.rint(nii.dataobj, dtype=data_dtype)
                        else:
                            new_data = data_dtype(nii.dataobj)
```

Comment (mgxd) [reply niworkflows/interfaces/bids.py:605]: nice catch!

File: niworkflows/interfaces/bids.py [hunk 3]
```diff
@@ -596,8 +596,19 @@ def _run_interface(self, runtime):
...
                         nii = nb.load(out_file, mmap=False)
                     data_dtype = np.dtype(data_dtype)
-                    if nii.get_data_dtype() != data_dtype:
+                    orig_dtype = nii.get_data_dtype()
+                    if orig_dtype != data_dtype:
+                        LOGGER.warning(
+                            f"Changing {out_file} dtype from {orig_dtype} to {data_dtype}"
+                        )
+                        # coerce dataobj to new data dtype
+                        if issubclass(data_dtype, np.integer):
```
Comment (oesteban) [inline niworkflows/interfaces/bids.py:605]:
```suggestion
                        if np.issubdtype(data_dtype, np.integer):
```

**Comment by mgxd** [comment]
I kept running into:
```
TypeError: No loop matching the specified signature and casting was found for ufunc rint
```
when trying to set `dtype` in the `rint()` call. 

Logically, it would have to round first and then apply the dtype to the array anyways.

**Comment by mgxd** [comment]
i'll merge this in now and cut a new release, thank you both for the reviews

**Pull request merged**